// @gentrace/core/models/expanded-test-run-all-of.ts
/* tslint:disable */
/* eslint-disable */
/**
 * Gentrace API
 * These API routes are designed to ingest events from clients.
 *
 * The version of the OpenAPI document: 0.22.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import { FullRun } from "./full-run";
// May contain unused imports in some cases
// @ts-ignore
import { ResolvedStepRun } from "./resolved-step-run";
// May contain unused imports in some cases
// @ts-ignore
import { TestCase } from "./test-case";
// May contain unused imports in some cases
// @ts-ignore
import { TestEvaluation } from "./test-evaluation";

/**
 *
 * @export
 * @interface ExpandedTestRunAllOf
 */
export interface ExpandedTestRunAllOf {
  /**
   *
   * @type {FullRun}
   * @memberof ExpandedTestRunAllOf
   */
  full?: FullRun;
  /**
   *
   * @type {Array<ResolvedStepRun>}
   * @memberof ExpandedTestRunAllOf
   */
  steps?: Array<ResolvedStepRun>;
  /**
   *
   * @type {TestCase}
   * @memberof ExpandedTestRunAllOf
   */
  case?: TestCase;
  /**
   *
   * @type {Array<TestEvaluation>}
   * @memberof ExpandedTestRunAllOf
   */
  evaluations?: Array<TestEvaluation>;
}
