// @gentrace/core/models/pipeline.ts
/* tslint:disable */
/* eslint-disable */
/**
 * Gentrace API
 * These API routes are designed to ingest events from clients.
 *
 * The version of the OpenAPI document: 0.22.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

/**
 *
 * @export
 * @interface Pipeline
 */
export interface Pipeline {
  /**
   * The ID of the pipeline
   * @type {string}
   * @memberof Pipeline
   */
  id: string;
  /**
   * The date and time when the pipeline was created
   * @type {string}
   * @memberof Pipeline
   */
  createdAt: string;
  /**
   * The date and time when the pipeline was last updated
   * @type {string}
   * @memberof Pipeline
   */
  updatedAt: string;
  /**
   * The date and time when the pipeline was archived, can be null if the pipeline has not been archived
   * @type {string}
   * @memberof Pipeline
   */
  archivedAt?: string | null;
  /**
   * The labels attached to the pipeline
   * @type {Array<string>}
   * @memberof Pipeline
   */
  labels: Array<string>;
  /**
   * The name of the pipeline
   * @type {string}
   * @memberof Pipeline
   */
  displayName?: string | null;
  /**
   * The slug of the pipeline
   * @type {string}
   * @memberof Pipeline
   */
  slug: string;
  /**
   * The ID of the organization that owns the pipeline
   * @type {string}
   * @memberof Pipeline
   */
  organizationId: string;
  /**
   * The branch that the pipeline is associated with
   * @type {string}
   * @memberof Pipeline
   */
  branch?: string | null;
}
