// @ironclad/rivet-core/dist/esm/plugins/openai/nodes/GetAssistantNode.js
import {} from '../../../index.js';
import { newId, getInputOrData } from '../../../utils/index.js';
import { pluginNodeDefinition } from '../../../model/NodeDefinition.js';
import { handleOpenAIError } from '../handleOpenaiError.js';
export const GetAssistantNodeImpl = {
    create() {
        return {
            id: newId(),
            type: 'openaiGetAssistant',
            data: {
                assistantId: '',
                useAssistantIdInput: true,
            },
            title: 'Get Assistant',
            visualData: {
                x: 0,
                y: 0,
                width: 300,
            },
        };
    },
    getUIData() {
        return {
            group: 'OpenAI',
            contextMenuTitle: 'Get Assistant',
            infoBoxTitle: 'Get Assistant Node',
            infoBoxBody: 'Retrieve an assistant by its ID from OpenAI.',
        };
    },
    getInputDefinitions(data) {
        const inputs = [];
        if (data.useAssistantIdInput) {
            inputs.push({
                id: 'assistantId',
                dataType: 'string',
                title: 'Assistant ID',
                coerced: true,
                defaultValue: '',
                description: 'The ID of the assistant to retrieve.',
                required: true,
            });
        }
        return inputs;
    },
    getOutputDefinitions() {
        return [
            {
                id: 'assistantId',
                dataType: 'string',
                title: 'Assistant ID',
                description: 'The ID of the retrieved assistant.',
            },
            {
                id: 'assistant',
                dataType: 'object',
                title: 'Assistant',
                description: 'The full retrieved assistant object.',
            },
        ];
    },
    getEditors() {
        return [
            {
                type: 'string',
                dataKey: 'assistantId',
                useInputToggleDataKey: 'useAssistantIdInput',
                label: 'Assistant ID',
                placeholder: 'Enter assistant ID',
            },
        ];
    },
    getBody(data) {
        return `Assistant ID: ${data.useAssistantIdInput ? '(Assistant ID From Input)' : data.assistantId}`;
    },
    async process(data, inputData, context) {
        const assistantId = getInputOrData(data, inputData, 'assistantId');
        if (!assistantId?.trim()) {
            throw new Error('Assistant ID is required.');
        }
        if (!context.settings.openAiKey) {
            throw new Error('OpenAI key is not set.');
        }
        const response = await fetch(`https://api.openai.com/v1/assistants/${assistantId}`, {
            method: 'GET',
            headers: {
                'OpenAI-Beta': 'assistants=v1',
                'Content-Type': 'application/json',
                Authorization: `Bearer ${context.settings.openAiKey}`,
                'OpenAI-Organization': context.settings.openAiOrganization ?? '',
            },
        });
        if (response.status === 404) {
            return {
                ['assistantId']: {
                    type: 'control-flow-excluded',
                    value: undefined,
                },
                ['assistant']: {
                    type: 'control-flow-excluded',
                    value: undefined,
                },
            };
        }
        await handleOpenAIError(response);
        const body = await response.json();
        return {
            ['assistantId']: {
                type: 'string',
                value: body.id,
            },
            ['assistant']: {
                type: 'object',
                value: body,
            },
        };
    },
};
export const getAssistantNode = pluginNodeDefinition(GetAssistantNodeImpl, 'Get Assistant');
