// @ironclad/rivet-core/dist/esm/plugins/openai/nodes/GetOpenAIFileNode.js
import {} from '../../../index.js';
import { newId, getInputOrData, dedent } from '../../../utils/index.js';
import { pluginNodeDefinition } from '../../../model/NodeDefinition.js';
import { handleOpenAIError } from '../handleOpenaiError.js';
export const GetOpenAIFileNodeImpl = {
    create() {
        return {
            id: newId(),
            type: 'openaiGetFile',
            data: {
                useFileIdInput: true,
            },
            title: 'Get OpenAI File',
            visualData: {
                x: 0,
                y: 0,
                width: 300,
            },
        };
    },
    getUIData() {
        return {
            group: 'OpenAI',
            contextMenuTitle: 'Get OpenAI File',
            infoBoxTitle: 'Get OpenAI File Node',
            infoBoxBody: 'Get a specific file from OpenAI.',
        };
    },
    getInputDefinitions(data) {
        const inputs = [];
        if (data.useFileIdInput) {
            inputs.push({
                id: 'fileId',
                dataType: 'string',
                title: 'File ID',
                coerced: true,
                defaultValue: '',
                description: 'The ID of the file to retrieve.',
                required: true,
            });
        }
        return inputs;
    },
    getOutputDefinitions() {
        return [
            {
                id: 'file',
                dataType: 'object',
                title: 'File',
                description: 'The retrieved file object.',
            },
        ];
    },
    getEditors() {
        return [
            {
                type: 'string',
                dataKey: 'fileId',
                useInputToggleDataKey: 'useFileIdInput',
                label: 'File ID',
            },
        ];
    },
    getBody(data) {
        return dedent `
      File ID: ${data.fileId ?? 'Unknown'}
    `;
    },
    async process(data, inputData, context) {
        const fileId = getInputOrData(data, inputData, 'fileId');
        if (!context.settings.openAiKey) {
            throw new Error('OpenAI key is not set.');
        }
        const url = `https://api.openai.com/v1/files/${fileId}`;
        const response = await fetch(url, {
            method: 'GET',
            headers: {
                Authorization: `Bearer ${context.settings.openAiKey}`,
                'OpenAI-Organization': context.settings.openAiOrganization ?? '',
            },
        });
        if (response.status === 404) {
            return {
                ['file']: {
                    type: 'control-flow-excluded',
                    value: undefined,
                },
            };
        }
        await handleOpenAIError(response);
        const body = await response.json();
        return {
            ['file']: {
                type: 'object',
                value: body,
            },
        };
    },
};
export const getOpenAIFileNode = pluginNodeDefinition(GetOpenAIFileNodeImpl, 'Get OpenAI File');
