// @ironclad/rivet-core/dist/esm/plugins/assemblyAi/LemurTaskNode.js
import { nanoid } from 'nanoid/non-secure';
import { dedent } from 'ts-dedent';
import { AssemblyAI } from 'assemblyai';
import {} from '../../index.js';
import { getApiKey, getLemurParams, lemurEditorDefinitions, lemurTranscriptIdsInputDefinition, } from './lemurHelpers.js';
import { pluginNodeDefinition } from '../../model/NodeDefinition.js';
import { coerceTypeOptional } from '../../utils/coerceType.js';
export const LemurTaskNodeImpl = {
    create() {
        const chartNode = {
            type: 'assemblyAiLemurTask',
            title: 'LeMUR Task',
            id: nanoid(),
            visualData: {
                x: 0,
                y: 0,
                width: 250,
            },
            data: {
                final_model: 'default',
            },
        };
        return chartNode;
    },
    getInputDefinitions() {
        return [
            lemurTranscriptIdsInputDefinition,
            {
                id: 'prompt',
                dataType: 'string',
                title: 'Prompt',
            },
        ];
    },
    getOutputDefinitions() {
        return [
            {
                dataType: 'string',
                id: 'response',
                title: 'Response',
            },
        ];
    },
    getEditors() {
        return [
            {
                type: 'string',
                label: 'Prompt',
                dataKey: 'prompt',
            },
            ...lemurEditorDefinitions,
        ];
    },
    getBody() {
        return '';
    },
    getUIData() {
        return {
            infoBoxBody: dedent `Use AssemblyAI LeMUR Custom Task to ask anything.`,
            infoBoxTitle: 'Use AssemblyAI LeMUR Custom Task',
            contextMenuTitle: 'LeMUR Custom Task',
            group: ['AI', 'AssemblyAI'],
        };
    },
    async process(data, inputs, context) {
        const apiKey = getApiKey(context);
        const client = new AssemblyAI({ apiKey });
        const params = {
            prompt: coerceTypeOptional(inputs['prompt'], 'string') || data.prompt || '',
            ...getLemurParams(inputs, data),
        };
        if (!params.prompt)
            throw new Error('Prompt must be provided.');
        const { response } = await client.lemur.task(params);
        return {
            ['response']: {
                type: 'string',
                value: response,
            },
        };
    },
};
export const lemurTaskNode = pluginNodeDefinition(LemurTaskNodeImpl, 'LeMUR Task');
