// @ironclad/rivet-core/dist/esm/plugins/assemblyAi/LemurActionItemsNode.js
import { nanoid } from 'nanoid/non-secure';
import { dedent } from 'ts-dedent';
import { AssemblyAI } from 'assemblyai';
import {} from '../../index.js';
import { pluginNodeDefinition } from '../../model/NodeDefinition.js';
import { getApiKey, getLemurParams, lemurEditorDefinitions, lemurTranscriptIdsInputDefinition, } from './lemurHelpers.js';
export const LemurActionItemsNodeImpl = {
    create() {
        const chartNode = {
            type: 'assemblyAiLemurActionItems',
            title: 'LeMUR Action Items',
            id: nanoid(),
            visualData: {
                x: 0,
                y: 0,
                width: 250,
            },
            data: {
                final_model: 'default',
            },
        };
        return chartNode;
    },
    getInputDefinitions() {
        return [
            lemurTranscriptIdsInputDefinition,
            {
                id: 'context',
                dataType: 'string',
                title: 'Context',
            },
        ];
    },
    getOutputDefinitions() {
        return [
            {
                dataType: 'string',
                id: 'response',
                title: 'Response',
            },
        ];
    },
    getEditors() {
        return [
            {
                type: 'string',
                label: 'Context',
                dataKey: 'context',
            },
            ...lemurEditorDefinitions,
        ];
    },
    getBody() {
        return '';
    },
    getUIData() {
        return {
            infoBoxBody: dedent `Use AssemblyAI LeMUR Action Items to extract action items`,
            infoBoxTitle: 'Use AssemblyAI LeMUR Action Items',
            contextMenuTitle: 'LeMUR Action Items',
            group: ['AI', 'AssemblyAI'],
        };
    },
    async process(data, inputs, context) {
        const apiKey = getApiKey(context);
        const client = new AssemblyAI({ apiKey });
        const params = getLemurParams(inputs, data);
        const { response } = await client.lemur.actionItems(params);
        return {
            ['response']: {
                type: 'string',
                value: response,
            },
        };
    },
};
export const lemurActionItemsNode = pluginNodeDefinition(LemurActionItemsNodeImpl, 'LeMUR Action Items');
