// @ironclad/rivet-core/dist/esm/model/nodes/AbortGraphNode.js
import { nanoid } from 'nanoid/non-secure';
import { NodeImpl } from '../NodeImpl.js';
import { dedent } from 'ts-dedent';
import { nodeDefinition } from '../NodeDefinition.js';
import { coerceTypeOptional } from '../../utils/coerceType.js';
export class AbortGraphNodeImpl extends NodeImpl {
    static create() {
        const chartNode = {
            type: 'abortGraph',
            title: 'Abort Graph',
            id: nanoid(),
            visualData: {
                x: 0,
                y: 0,
                width: 200,
            },
            data: {
                successfully: true,
                errorMessage: '',
            },
        };
        return chartNode;
    }
    getInputDefinitions() {
        const inputs = [
            {
                id: 'data',
                title: 'Data or Error',
                dataType: 'any',
                description: 'The message to abort the graph with.',
            },
        ];
        if (this.data.useSuccessfullyInput) {
            inputs.push({
                id: 'successfully',
                title: 'Successfully',
                dataType: 'boolean',
                description: 'Whether to successfully abort the graph (early-exit), or error abort the graph.',
            });
        }
        return inputs;
    }
    getOutputDefinitions() {
        return [];
    }
    getEditors() {
        return [
            {
                type: 'toggle',
                label: 'Successfully Abort',
                dataKey: 'successfully',
                useInputToggleDataKey: 'useSuccessfullyInput',
            },
            {
                type: 'string',
                label: 'Error Message (if not successfully aborting)',
                dataKey: 'errorMessage',
            },
        ];
    }
    getBody() {
        return dedent `
      ${this.data.useSuccessfullyInput
            ? 'Success depends on input'
            : this.data.successfully
                ? 'Successfully Abort'
                : this.data.errorMessage
                    ? `Error Abort: ${this.data.errorMessage}`
                    : 'Error Abort'}
    `;
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Aborts the execution of the entire graph immediately.

        Can either "successfully" abort the graph (early-exit), or "error" abort the graph.
      `,
            infoBoxTitle: 'Abort Graph Node',
            contextMenuTitle: 'Abort Graph',
            group: ['Logic'],
        };
    }
    async process(inputs, context) {
        const successfully = this.data.useSuccessfullyInput
            ? coerceTypeOptional(inputs['successfully'], 'boolean') ?? this.data.successfully
            : this.data.successfully;
        if (successfully) {
            context.abortGraph();
        }
        else {
            const errorMessage = coerceTypeOptional(inputs['data'], 'string')?.trim() ||
                this.data.errorMessage ||
                'Graph aborted with error';
            context.abortGraph(errorMessage);
        }
        return {};
    }
}
export const abortGraphNode = nodeDefinition(AbortGraphNodeImpl, 'Abort Graph');
