// @ironclad/rivet-core/dist/esm/model/nodes/ContextNode.js
import {} from '../NodeBase.js';
import { nanoid } from 'nanoid/non-secure';
import { NodeImpl } from '../NodeImpl.js';
import {} from '../DataValue.js';
import {} from '../GraphProcessor.js';
import {} from '../ProcessContext.js';
import { dedent } from 'ts-dedent';
import {} from '../EditorDefinition.js';
import {} from '../NodeBodySpec.js';
import { nodeDefinition } from '../NodeDefinition.js';
export class ContextNodeImpl extends NodeImpl {
    static create() {
        const chartNode = {
            type: 'context',
            title: 'Context',
            id: nanoid(),
            visualData: {
                x: 0,
                y: 0,
                width: 300,
            },
            data: {
                id: 'input',
                dataType: 'string',
                defaultValue: undefined,
                useDefaultValueInput: false,
            },
        };
        return chartNode;
    }
    getInputDefinitions() {
        if (this.data.useDefaultValueInput) {
            return [
                {
                    id: 'default',
                    title: 'Default Value',
                    dataType: this.chartNode.data.dataType,
                },
            ];
        }
        return [];
    }
    getOutputDefinitions() {
        return [
            {
                id: 'data',
                title: this.data.id,
                dataType: this.chartNode.data.dataType,
            },
        ];
    }
    getEditors() {
        return [
            { type: 'string', label: 'ID', dataKey: 'id' },
            { type: 'dataTypeSelector', label: 'Data Type', dataKey: 'dataType' },
            {
                type: 'anyData',
                label: 'Default Value',
                dataKey: 'defaultValue',
                useInputToggleDataKey: 'useDefaultValueInput',
            },
        ];
    }
    getBody() {
        return dedent `
      ${this.data.id}
      Type: ${this.data.dataType}
    `;
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Retrieves a value from the graph's context using a configured id. The context serves as a "global graph input", allowing the same values to be accessible from any graph or subgraph.
      `,
            infoBoxTitle: 'Context Node',
            contextMenuTitle: 'Context',
            group: ['Advanced'],
        };
    }
    async process(inputs, context) {
        const contextValue = context.contextValues[this.data.id];
        if (contextValue !== undefined) {
            return {
                ['data']: contextValue,
            };
        }
        let defaultValue;
        if (this.data.useDefaultValueInput) {
            defaultValue = inputs['default'];
        }
        else {
            defaultValue = { type: this.data.dataType, value: this.data.defaultValue };
        }
        return {
            ['data']: defaultValue,
        };
    }
}
export const contextNode = nodeDefinition(ContextNodeImpl, 'Context');
