// @ironclad/rivet-core/dist/esm/model/nodes/FilterNode.js
import {} from '../NodeBase.js';
import { nanoid } from 'nanoid/non-secure';
import { NodeImpl } from '../NodeImpl.js';
import { nodeDefinition } from '../NodeDefinition.js';
import {} from '../GraphProcessor.js';
import {} from '../../index.js';
import { zip } from 'lodash-es';
import { dedent } from 'ts-dedent';
import { coerceType } from '../../utils/coerceType.js';
export class FilterNodeImpl extends NodeImpl {
    static create() {
        const chartNode = {
            type: 'filter',
            title: 'Filter',
            id: nanoid(),
            visualData: {
                x: 0,
                y: 0,
                width: 175,
            },
            data: {},
        };
        return chartNode;
    }
    getInputDefinitions() {
        return [
            {
                id: 'array',
                title: 'Array',
                dataType: 'any[]',
                required: true,
            },
            {
                id: 'include',
                title: 'Include',
                dataType: 'boolean[]',
                required: true,
            },
        ];
    }
    getOutputDefinitions() {
        return [
            {
                id: 'filtered',
                title: 'Filtered',
                dataType: 'any[]',
            },
        ];
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Takes in both an array of values, and an array of booleans of the same length, and filters the array where the corresponding boolean is true.
      `,
            infoBoxTitle: 'Filter Node',
            contextMenuTitle: 'Filter',
            group: ['Lists'],
        };
    }
    async process(inputs) {
        const array = coerceType(inputs['array'], 'any[]');
        const include = coerceType(inputs['include'], 'boolean[]');
        const zipped = zip(array, include);
        const filtered = zipped.filter(([_, include]) => include).map(([value, _]) => value);
        return {
            ['filtered']: {
                type: inputs['array']?.type ?? 'any',
                value: filtered,
            },
        };
    }
}
export const filterNode = nodeDefinition(FilterNodeImpl, 'Filter');
