// @ironclad/rivet-core/dist/esm/model/nodes/HashNode.js
import {} from '../NodeBase.js';
import { nanoid } from 'nanoid/non-secure';
import { NodeImpl } from '../NodeImpl.js';
import { nodeDefinition } from '../NodeDefinition.js';
import {} from '../../index.js';
import * as crypto from 'crypto-js';
import { match } from 'ts-pattern';
import { dedent } from 'ts-dedent';
import { coerceType } from '../../utils/coerceType.js';
const { SHA256, SHA512, MD5, SHA1 } = crypto;
export class HashNodeImpl extends NodeImpl {
    static create() {
        const chartNode = {
            type: 'hash',
            title: 'Hash',
            id: nanoid(),
            visualData: {
                x: 0,
                y: 0,
                width: 250,
            },
            data: {
                algorithm: 'sha256',
            },
        };
        return chartNode;
    }
    getInputDefinitions() {
        return [
            {
                id: 'input',
                title: 'Input',
                dataType: 'string',
                required: true,
            },
        ];
    }
    getOutputDefinitions() {
        return [
            {
                id: 'hash',
                title: 'Hash',
                dataType: 'string',
            },
        ];
    }
    getEditors() {
        return [
            {
                type: 'dropdown',
                label: 'Algorithm',
                dataKey: 'algorithm',
                options: [
                    { value: 'md5', label: 'MD5' },
                    { value: 'sha1', label: 'SHA1' },
                    { value: 'sha256', label: 'SHA256' },
                    { value: 'sha512', label: 'SHA512' },
                ],
            },
        ];
    }
    getBody() {
        return algorithmDisplayName[this.data.algorithm];
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Computes a hash of the input value using the configured hash function.
      `,
            infoBoxTitle: 'Hash Node',
            contextMenuTitle: 'Hash',
            group: ['Data'],
        };
    }
    async process(inputs) {
        const inputText = coerceType(inputs['input'], 'string');
        const hash = match(this.data.algorithm)
            .with('md5', () => MD5(inputText).toString())
            .with('sha1', () => SHA1(inputText).toString())
            .with('sha256', () => SHA256(inputText).toString())
            .with('sha512', () => SHA512(inputText).toString())
            .exhaustive();
        return {
            ['hash']: {
                type: 'string',
                value: hash,
            },
        };
    }
}
const algorithmDisplayName = {
    md5: 'MD5',
    sha1: 'SHA-1',
    sha256: 'SHA-256',
    sha512: 'SHA-512',
};
export const hashNode = nodeDefinition(HashNodeImpl, 'Hash');
