// @ironclad/rivet-core/dist/esm/model/nodes/IfElseNode.js
import { NodeImpl } from '../NodeImpl.js';
import { nodeDefinition } from '../NodeDefinition.js';
import {} from '../NodeBase.js';
import {} from '../DataValue.js';
import { nanoid } from 'nanoid/non-secure';
import { dedent } from 'ts-dedent';
import { coerceType } from '../../utils/coerceType.js';
export class IfElseNodeImpl extends NodeImpl {
    static create = () => {
        const chartNode = {
            type: 'ifElse',
            title: 'If/Else',
            id: nanoid(),
            data: {
                // Legacy behavior is false
                unconnectedControlFlowExcluded: true,
            },
            visualData: {
                x: 0,
                y: 0,
                width: 175,
            },
        };
        return chartNode;
    };
    getInputDefinitions() {
        return [
            {
                id: 'if',
                title: 'If',
                dataType: 'any',
                description: 'If this is truthy, the `true` value will be passed through the output port. Otherwise, the `false` value will be passed through the output port. An unconnected port is considered false. A `Not Ran` value is considered false.',
            },
            {
                id: 'true',
                title: 'True',
                dataType: 'any',
                description: 'The value to pass through the output port if the condition is truthy. ',
            },
            {
                id: 'false',
                title: 'False',
                dataType: 'any',
                description: 'The value to pass through the output port if the condition is not truthy.',
            },
        ];
    }
    getOutputDefinitions() {
        return [
            {
                id: 'output',
                title: 'Output',
                dataType: 'any',
                description: 'The `true` or `false` value, depending on the `if` condition.',
            },
        ];
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Takes in three inputs: a condition, a true value, and a false value. If the condition is truthy, the true value is passed through the output port. If the condition is not truthy, the false value is passed through the output port.

        This node can "consume" a \`Not Ran\` to continue a graph from that point.
      `,
            infoBoxTitle: 'If/Else Node',
            contextMenuTitle: 'If/Else',
            group: ['Logic'],
        };
    }
    getEditors() {
        return [
            {
                type: 'toggle',
                label: "Don't run unconnected ports",
                dataKey: 'unconnectedControlFlowExcluded',
            },
        ];
    }
    async process(inputData) {
        const unconnectedValue = this.data.unconnectedControlFlowExcluded
            ? { type: 'control-flow-excluded', value: undefined }
            : {
                type: 'any',
                value: undefined,
            };
        const ifValue = inputData['if'];
        const trueValue = inputData['true'] ?? unconnectedValue;
        const falseValue = inputData['false'] ?? unconnectedValue;
        if (!(trueValue || falseValue)) {
            return {
                ['output']: {
                    type: 'control-flow-excluded',
                    value: undefined,
                },
            };
        }
        if (ifValue?.type === 'control-flow-excluded') {
            return {
                ['output']: falseValue,
            };
        }
        if (ifValue?.value == null) {
            return {
                ['output']: falseValue,
            };
        }
        if (ifValue?.type && ifValue.type === 'boolean') {
            return {
                ['output']: ifValue.value ? trueValue : falseValue,
            };
        }
        if (ifValue?.type === 'string') {
            return {
                ['output']: ifValue.value.length > 0 ? trueValue : falseValue,
            };
        }
        if (ifValue?.type === 'chat-message') {
            const asString = coerceType(ifValue, 'string');
            return {
                ['output']: asString ? trueValue : falseValue,
            };
        }
        if (ifValue?.type.endsWith('[]')) {
            return {
                ['output']: ifValue.value.length > 0 ? trueValue : falseValue,
            };
        }
        if (ifValue?.type === 'any' || ifValue?.type === 'object') {
            return {
                ['output']: !!ifValue.value ? trueValue : falseValue,
            };
        }
        return {
            ['output']: falseValue,
        };
    }
}
export const ifElseNode = nodeDefinition(IfElseNodeImpl, 'If/Else');
