// @ironclad/rivet-core/dist/esm/model/nodes/IfNode.js
import { NodeImpl } from '../NodeImpl.js';
import { nodeDefinition } from '../NodeDefinition.js';
import {} from '../NodeBase.js';
import {} from '../DataValue.js';
import { nanoid } from 'nanoid/non-secure';
import { dedent } from 'ts-dedent';
import { coerceType } from '../../utils/coerceType.js';
export class IfNodeImpl extends NodeImpl {
    static create = () => {
        const chartNode = {
            type: 'if',
            title: 'If',
            id: nanoid(),
            data: {
                // Legacy behavior is false
                unconnectedControlFlowExcluded: true,
            },
            visualData: {
                x: 0,
                y: 0,
                width: 125,
            },
        };
        return chartNode;
    };
    getInputDefinitions() {
        return [
            {
                id: 'if',
                title: 'If',
                dataType: 'any',
                description: 'If this is truthy, the value will be passed through the True port. Otherwise, it will be passed through the False port. An unconnected port is considered false.',
            },
            {
                id: 'value',
                title: 'Value',
                dataType: 'any',
                description: 'The value to pass through the True or False port. If unconnected, it will be undefined.',
            },
        ];
    }
    getOutputDefinitions() {
        return [
            {
                id: 'output',
                title: 'True',
                dataType: 'any',
                description: 'The `value` passed through if the condition is truthy.',
            },
            {
                id: 'falseOutput',
                title: 'False',
                dataType: 'any',
                description: 'The `value` passed through if the condition is falsy.',
            },
        ];
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Takes in a condition and a value. If the condition is truthy, the value is passed through the True port, and the False port is not ran.
        If the condition is falsy, the value is passed through the False port, and the True port is not ran.
      `,
            infoBoxTitle: 'If Node',
            contextMenuTitle: 'If',
            group: ['Logic'],
        };
    }
    getEditors() {
        return [
            {
                type: 'toggle',
                label: "Don't run unconnected value",
                dataKey: 'unconnectedControlFlowExcluded',
            },
        ];
    }
    async process(inputData) {
        const unconnectedValue = this.data.unconnectedControlFlowExcluded
            ? { type: 'control-flow-excluded', value: undefined }
            : { type: 'any', value: undefined };
        const ifValue = inputData['if'];
        const value = inputData['value'] ?? unconnectedValue;
        const isFalse = {
            output: {
                type: 'control-flow-excluded',
                value: undefined,
            },
            falseOutput: value,
        };
        if (!ifValue) {
            return isFalse;
        }
        if (ifValue.type === 'control-flow-excluded') {
            return isFalse;
        }
        if (ifValue.type === 'string' && !ifValue.value) {
            return isFalse;
        }
        if (ifValue.type === 'boolean' && !ifValue.value) {
            return isFalse;
        }
        if (ifValue.type === 'chat-message') {
            const asString = coerceType(ifValue, 'string');
            if (!asString) {
                return isFalse;
            }
        }
        if (ifValue.type.endsWith('[]')) {
            const value = ifValue;
            if (!value.value || value.value.length === 0) {
                return isFalse;
            }
        }
        return {
            ['output']: value,
            ['falseOutput']: {
                type: 'control-flow-excluded',
                value: undefined,
            },
        };
    }
}
export const ifNode = nodeDefinition(IfNodeImpl, 'If');
