// @ironclad/rivet-core/dist/esm/model/nodes/LoopControllerNode.js
import {} from '../NodeBase.js';
import { nanoid } from 'nanoid/non-secure';
import { NodeImpl } from '../NodeImpl.js';
import { nodeDefinition } from '../NodeDefinition.js';
import {} from '../GraphProcessor.js';
import { coerceType } from '../../utils/coerceType.js';
import {} from '../ProcessContext.js';
import { dedent } from 'ts-dedent';
import {} from '../../index.js';
import { entries, values } from '../../utils/typeSafety.js';
export class LoopControllerNodeImpl extends NodeImpl {
    static create() {
        const chartNode = {
            type: 'loopController',
            title: 'Loop Controller',
            id: nanoid(),
            visualData: {
                x: 0,
                y: 0,
                width: 250,
            },
            data: {
                maxIterations: 100,
            },
        };
        return chartNode;
    }
    getInputDefinitions(connections, nodes) {
        const inputs = [];
        const messageCount = this.#getInputPortCount(connections);
        inputs.push({
            dataType: 'any',
            id: 'continue',
            title: 'Continue',
        });
        let i = 1;
        for (; i <= messageCount + 1; i++) {
            const input = {
                dataType: 'any',
                id: `input${i}`,
                title: `Input ${i}`,
            };
            const inputConnection = connections.find((connection) => connection.inputId === input.id && connection.inputNodeId === this.id);
            if (inputConnection && nodes[inputConnection.outputNodeId]) {
                input.title = nodes[inputConnection.outputNodeId].title;
            }
            const inputDefault = {
                dataType: 'any',
                id: `input${i}Default`,
                title: `Input ${i} Default`,
            };
            const inputDefaultConnection = connections.find((connection) => connection.inputId === inputDefault.id && connection.inputNodeId === this.id);
            if (inputDefaultConnection && nodes[inputDefaultConnection.outputNodeId]) {
                inputDefault.title = `${nodes[inputDefaultConnection.outputNodeId].title} (Default)`;
            }
            inputs.push(input);
            inputs.push(inputDefault);
        }
        return inputs;
    }
    getOutputDefinitions(connections, nodes) {
        const messageCount = this.#getInputPortCount(connections);
        const outputs = [];
        outputs.push({
            dataType: 'any',
            id: 'break',
            title: 'Break',
        });
        outputs.push({
            dataType: 'number',
            id: 'iteration',
            title: 'Iteration',
        });
        for (let i = 1; i <= messageCount; i++) {
            const output = {
                dataType: 'any',
                id: `output${i}`,
                title: `Output ${i}`,
            };
            const inputConnection = connections.find((connection) => connection.inputId === `input${i}` && connection.inputNodeId === this.id);
            if (inputConnection && nodes[inputConnection.outputNodeId]) {
                output.title = `${nodes[inputConnection.outputNodeId].title}?`;
            }
            outputs.push(output);
        }
        return outputs;
    }
    getEditors() {
        return [
            {
                type: 'number',
                label: 'Max Iterations',
                dataKey: 'maxIterations',
            },
            {
                type: 'dropdown',
                options: [
                    {
                        label: 'Break',
                        value: 'break',
                    },
                    {
                        label: 'Error',
                        value: 'error',
                    },
                ],
                label: 'At Max Iterations',
                dataKey: 'atMaxIterationsAction',
                defaultValue: 'error',
                helperMessage: 'What should happen when the max iterations is reached?',
            },
        ];
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Defines the entry point for a loop. Values from inside the loop should be passed back through the "Input" ports, and their corresponding "Default" values can be specified on the input ports as well.

        If the "continue" input is falsey, then the "break" output will run.
      `,
            infoBoxTitle: 'Loop Controller Node',
            contextMenuTitle: 'Loop Controller',
            group: ['Logic'],
        };
    }
    #getInputPortCount(connections) {
        const inputNodeId = this.chartNode.id;
        const messageConnections = connections.filter((connection) => connection.inputNodeId === inputNodeId && connection.inputId.startsWith('input'));
        let maxMessageNumber = 0;
        for (const connection of messageConnections) {
            const messageNumber = parseInt(connection.inputId.replace('input', ''));
            if (messageNumber > maxMessageNumber) {
                maxMessageNumber = messageNumber;
            }
        }
        return maxMessageNumber;
    }
    async process(inputs, context) {
        const output = {};
        let inputCount = 0;
        while (inputs[`input${inputCount + 1}`] || inputs[`input${inputCount + 1}Default`]) {
            inputCount++;
        }
        const defaultInputs = entries(inputs).filter(([key]) => key.endsWith('Default'));
        // If any of the default inputs are control-flow-excluded, then exclude all outputs.
        // Technically, it should be "any node outside of the cycle" but this should be good enough?
        if (defaultInputs.some(([, value]) => value?.type === 'control-flow-excluded')) {
            for (let i = 0; i <= inputCount; i++) {
                output[`output${i}`] = { type: 'control-flow-excluded', value: undefined };
            }
            output['break'] = { type: 'control-flow-excluded', value: undefined };
            return output;
        }
        const iterationCount = context.attachedData.loopInfo?.iterationCount ?? 0;
        output['iteration'] = { type: 'number', value: iterationCount + 1 };
        if (iterationCount >= (this.data.maxIterations ?? 100) && this.data.atMaxIterationsAction !== 'break') {
            throw new Error(`Loop controller exceeded max iterations of ${this.data.maxIterations ?? 100}`);
        }
        // If the continue port is not connected (so undefined), or if it's undefined before it's
        // inside the loop itself (connection has not ran yet), then we should continue by default.
        let continueValue = false;
        if (inputs['continue'] === undefined) {
            continueValue = true;
        }
        else {
            const continueDataValue = inputs['continue'];
            if (continueDataValue.type === 'control-flow-excluded') {
                continueValue = false;
            }
            else {
                continueValue = coerceType(continueDataValue, 'boolean');
            }
        }
        if (iterationCount >= (this.data.maxIterations ?? 100) && this.data.atMaxIterationsAction === 'break') {
            continueValue = false;
        }
        if (continueValue) {
            output['break'] = { type: 'control-flow-excluded', value: 'loop-not-broken' };
        }
        else {
            const inputValues = [];
            for (let i = 1; i <= inputCount; i++) {
                inputValues.push(inputs[`input${i}`]?.value);
            }
            // Break gets an array of all the input values
            output['break'] = { type: 'any[]', value: inputValues };
        }
        for (let i = 1; i <= inputCount; i++) {
            if (continueValue) {
                const inputId = `input${i}`;
                const outputId = `output${i}`;
                if (inputs[inputId]) {
                    output[outputId] = inputs[inputId];
                }
                else {
                    output[outputId] = inputs[`${inputId}Default`];
                }
            }
            else {
                output[`output${i}`] = { type: 'control-flow-excluded', value: undefined };
            }
        }
        return output;
    }
}
export const loopControllerNode = nodeDefinition(LoopControllerNodeImpl, 'Loop Controller');
