// @ironclad/rivet-core/dist/esm/model/nodes/MatchNode.js
import {} from '../NodeBase.js';
import { nanoid } from 'nanoid/non-secure';
import { NodeImpl } from '../NodeImpl.js';
import { nodeDefinition } from '../NodeDefinition.js';
import {} from '../DataValue.js';
import {} from '../../index.js';
import { dedent } from 'ts-dedent';
import { coerceType } from '../../utils/coerceType.js';
export class MatchNodeImpl extends NodeImpl {
    static create() {
        const chartNode = {
            type: 'match',
            title: 'Match',
            id: nanoid(),
            visualData: {
                x: 0,
                y: 0,
                width: 250,
            },
            data: {
                cases: ['YES', 'NO'],
            },
        };
        return chartNode;
    }
    getInputDefinitions() {
        const inputs = [
            {
                id: 'input',
                title: 'Test',
                dataType: 'string',
                required: true,
                description: 'The value that will be tested against each of the cases.',
            },
            {
                id: 'value',
                title: 'Value',
                dataType: 'any',
                description: 'The value passed through to the output port that matches. If unconnected, the test value will be passed through.',
            },
        ];
        return inputs;
    }
    getOutputDefinitions() {
        const outputs = [];
        for (let i = 0; i < this.data.cases.length; i++) {
            outputs.push({
                id: `case${i + 1}`,
                title: this.data.cases[i]?.trim() ? this.data.cases[i] : `Case ${i + 1}`,
                dataType: 'string',
                description: `The 'value' (or 'test' if value is unconnected) passed through if the test value matches this regex: /${this
                    .data.cases[i]}/`,
            });
        }
        outputs.push({
            id: 'unmatched',
            title: 'Unmatched',
            dataType: 'string',
            description: 'The value (or test if value is unconnected) passed through if no regexes match.',
        });
        return outputs;
    }
    getBody() {
        return dedent `
      ${this.data.exclusive ? 'First Matching Case' : 'All Matching Cases'}
      ${this.data.cases.length} Cases
    `;
    }
    getEditors() {
        return [
            {
                type: 'toggle',
                dataKey: 'exclusive',
                label: 'Exclusive',
                helperMessage: 'If enabled, only the first matching branch will be ran.',
            },
            {
                type: 'stringList',
                dataKey: 'cases',
                label: 'Cases',
                placeholder: 'Case (regular expression)',
                helperMessage: '(Regular expressions)',
            },
        ];
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Any number of regular expressions can be configured, each corresponding to an output of the node. The output port of the first matching regex will be ran, and all other output ports will not be ran.
      `,
            infoBoxTitle: 'Match Node',
            contextMenuTitle: 'Match',
            group: ['Logic'],
        };
    }
    async process(inputs) {
        const inputString = coerceType(inputs['input'], 'string');
        const value = inputs['value'];
        const outputType = value === undefined ? 'string' : value.type;
        const outputValue = value === undefined ? inputString : value.value;
        const cases = this.data.cases;
        let matched = false;
        const output = {};
        for (let i = 0; i < cases.length; i++) {
            const regExp = new RegExp(cases[i]);
            const match = regExp.test(inputString);
            const canMatch = !this.data.exclusive || !matched;
            if (match && canMatch) {
                matched = true;
                output[`case${i + 1}`] = {
                    type: outputType,
                    value: outputValue,
                };
            }
            else {
                output[`case${i + 1}`] = {
                    type: 'control-flow-excluded',
                    value: undefined,
                };
            }
        }
        if (!matched) {
            output['unmatched'] = {
                type: outputType,
                value: outputValue,
            };
        }
        else {
            output['unmatched'] = {
                type: 'control-flow-excluded',
                value: undefined,
            };
        }
        return output;
    }
}
export const matchNode = nodeDefinition(MatchNodeImpl, 'Match');
