// @ironclad/rivet-core/dist/esm/model/nodes/RaiseEventNode.js
import {} from '../NodeBase.js';
import { NodeImpl } from '../NodeImpl.js';
import { nodeDefinition } from '../NodeDefinition.js';
import {} from '../DataValue.js';
import { nanoid } from 'nanoid/non-secure';
import {} from '../GraphProcessor.js';
import { coerceType } from '../../utils/coerceType.js';
import {} from '../ProcessContext.js';
import { dedent } from 'ts-dedent';
import {} from '../EditorDefinition.js';
import {} from '../NodeBodySpec.js';
export class RaiseEventNodeImpl extends NodeImpl {
    static create() {
        return {
            id: nanoid(),
            type: 'raiseEvent',
            title: 'Raise Event',
            visualData: { x: 0, y: 0, width: 150 },
            data: {
                eventName: 'toast',
                useEventNameInput: false,
            },
        };
    }
    getInputDefinitions() {
        const inputDefinitions = [];
        if (this.chartNode.data.useEventNameInput) {
            inputDefinitions.push({
                id: 'eventName',
                title: 'Event Name',
                dataType: 'string',
            });
        }
        inputDefinitions.push({
            id: 'data',
            title: 'Data',
            dataType: 'any',
        });
        return inputDefinitions;
    }
    getOutputDefinitions() {
        return [
            {
                id: 'result',
                title: 'Result',
                dataType: 'any',
            },
        ];
    }
    getEditors() {
        return [
            {
                type: 'string',
                label: 'Event Name',
                dataKey: 'eventName',
                useInputToggleDataKey: 'useEventNameInput',
            },
        ];
    }
    getBody() {
        return this.data.useEventNameInput ? '(Using Input)' : this.data.eventName;
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Raises an event that the host project or a 'Wait For Event' node can listen for.
      `,
            infoBoxTitle: 'Raise Event Node',
            contextMenuTitle: 'Raise Event',
            group: ['Advanced'],
        };
    }
    async process(inputs, context) {
        const eventName = this.chartNode.data.useEventNameInput
            ? coerceType(inputs['eventName'], 'string')
            : this.chartNode.data.eventName;
        const eventData = inputs['data'];
        context.raiseEvent(eventName, eventData);
        return {
            result: eventData,
        };
    }
}
export const raiseEventNode = nodeDefinition(RaiseEventNodeImpl, 'Raise Event');
