// @ironclad/rivet-core/dist/esm/model/nodes/SetGlobalNode.js
import {} from '../NodeBase.js';
import { nanoid } from 'nanoid/non-secure';
import { NodeImpl } from '../NodeImpl.js';
import { nodeDefinition } from '../NodeDefinition.js';
import { isArrayDataType, isScalarDataType, scalarDefaults, unwrapDataValue, } from '../DataValue.js';
import {} from '../GraphProcessor.js';
import { coerceType } from '../../utils/coerceType.js';
import {} from '../ProcessContext.js';
import { dedent } from 'ts-dedent';
import {} from '../EditorDefinition.js';
import {} from '../NodeBodySpec.js';
export class SetGlobalNodeImpl extends NodeImpl {
    static create() {
        const chartNode = {
            type: 'setGlobal',
            title: 'Set Global',
            id: nanoid(),
            visualData: {
                x: 0,
                y: 0,
                width: 200,
            },
            data: {
                id: 'variable-name',
                dataType: 'string',
                useIdInput: false,
            },
        };
        return chartNode;
    }
    getInputDefinitions() {
        const inputs = [
            {
                id: 'value',
                title: 'Value',
                dataType: this.chartNode.data.dataType,
            },
        ];
        if (this.data.useIdInput) {
            inputs.push({
                id: 'id',
                title: 'Variable ID',
                dataType: 'string',
            });
        }
        return inputs;
    }
    getOutputDefinitions() {
        return [
            {
                id: 'saved-value',
                title: 'Value',
                dataType: this.data.dataType,
            },
            {
                id: 'previous-value',
                title: 'Previous Value',
                dataType: this.data.dataType,
            },
            {
                id: 'variable_id_out',
                title: 'Variable ID',
                dataType: 'string',
            },
        ];
    }
    getEditors() {
        return [
            {
                type: 'string',
                dataKey: 'id',
                useInputToggleDataKey: 'useIdInput',
                label: 'ID',
            },
            {
                type: 'dataTypeSelector',
                dataKey: 'dataType',
                label: 'Data Type',
                useInputToggleDataKey: 'useIdInput',
            },
        ];
    }
    getBody() {
        return dedent `
      ${this.data.id}
      Type: ${this.data.dataType}
    `;
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Sets a global value that is shared across all graphs and subgraphs. The id of the global value and the value itself are configured in this node.
      `,
            infoBoxTitle: 'Set Global Node',
            contextMenuTitle: 'Set Global',
            group: ['Advanced'],
        };
    }
    async process(inputs, context) {
        const rawValue = inputs['value'];
        if (!rawValue) {
            return {};
        }
        const id = this.data.useIdInput ? coerceType(inputs['id'], 'string') : this.data.id;
        if (!id) {
            throw new Error('Missing variable ID');
        }
        let previousValue = context.getGlobal(this.data.id);
        if (!previousValue && isArrayDataType(this.data.dataType)) {
            previousValue = { type: this.data.dataType, value: [] };
        }
        else if (!previousValue && isScalarDataType(this.data.dataType)) {
            previousValue = { type: this.data.dataType, value: scalarDefaults[this.data.dataType] };
        }
        const value = unwrapDataValue(rawValue);
        context.setGlobal(id, value);
        return {
            ['saved-value']: value,
            ['previous-value']: previousValue,
            ['variable_id_out']: { type: 'string', value: id },
        };
    }
}
export const setGlobalNode = nodeDefinition(SetGlobalNodeImpl, 'Set Global');
