// @ironclad/rivet-core/dist/esm/model/nodes/SplitNode.js
import { NodeImpl } from '../NodeImpl.js';
import { coerceType, dedent, getInputOrData, handleEscapeCharacters, newId } from '../../utils/index.js';
import { nodeDefinition } from '../NodeDefinition.js';
export class SplitNodeImpl extends NodeImpl {
    static create() {
        return {
            id: newId(),
            type: 'split',
            title: 'Split Text',
            visualData: { x: 0, y: 0, width: 250 },
            data: {
                delimiter: ',',
                regex: false,
            },
        };
    }
    getInputDefinitions() {
        const inputs = [
            {
                id: 'string',
                title: 'String',
                dataType: 'string',
            },
        ];
        if (this.data.useDelimiterInput) {
            inputs.push({
                id: 'delimiter',
                title: 'Delimiter',
                dataType: 'string',
            });
        }
        return inputs;
    }
    getOutputDefinitions() {
        return [
            {
                id: 'splitString',
                title: 'Split',
                dataType: 'string[]',
            },
        ];
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Splits a string by the provided delimiter.
      `,
            infoBoxTitle: 'Split Text Node',
            contextMenuTitle: 'Split Text',
            group: ['Text'],
        };
    }
    getEditors() {
        return [
            {
                type: 'toggle',
                label: 'Regex',
                dataKey: 'regex',
            },
            {
                type: 'code',
                label: 'Delimiter',
                language: 'plaintext',
                dataKey: 'delimiter',
                useInputToggleDataKey: 'useDelimiterInput',
            },
        ];
    }
    getBody() {
        if (this.data.useDelimiterInput) {
            return '(Delimiter from input)';
        }
        const normalized = handleEscapeCharacters(this.data.delimiter);
        if (normalized === '\n') {
            return '(New line)';
        }
        if (normalized === '\r\n') {
            return '(New line (windows))';
        }
        if (normalized === '\t') {
            return '(Tab)';
        }
        if (normalized === ' ') {
            return '(Space)';
        }
        return normalized;
    }
    async process(inputs, context) {
        const delimiter = getInputOrData(this.data, inputs, 'delimiter');
        const normalizedDelimiter = this.data.regex ? new RegExp(delimiter) : handleEscapeCharacters(delimiter);
        const stringToSplit = coerceType(inputs['string'], 'string');
        const splitString = stringToSplit.split(normalizedDelimiter);
        return {
            ['splitString']: {
                type: 'string[]',
                value: splitString,
            },
        };
    }
}
export const splitNode = nodeDefinition(SplitNodeImpl, 'Split String');
