// @ironclad/rivet-core/dist/esm/model/nodes/ToJsonNode.js
import {} from '../NodeBase.js';
import { nanoid } from 'nanoid/non-secure';
import { NodeImpl } from '../NodeImpl.js';
import { nodeDefinition } from '../NodeDefinition.js';
import {} from '../GraphProcessor.js';
import {} from '../../index.js';
import { dedent } from 'ts-dedent';
import { coerceType } from '../../utils/coerceType.js';
export class ToJsonNodeImpl extends NodeImpl {
    static create() {
        const chartNode = {
            type: 'toJson',
            title: 'To JSON',
            id: nanoid(),
            visualData: {
                x: 0,
                y: 0,
                width: 175,
            },
            data: {
                indented: true,
            },
        };
        return chartNode;
    }
    getInputDefinitions() {
        return [
            {
                id: 'data',
                title: 'Data',
                dataType: 'any',
                required: true,
            },
        ];
    }
    getOutputDefinitions() {
        return [
            {
                id: 'json',
                title: 'JSON',
                dataType: 'string',
            },
        ];
    }
    getEditors() {
        return [
            {
                type: 'toggle',
                label: 'Indented',
                dataKey: 'indented',
            },
        ];
    }
    getBody() {
        return this.data.indented ? 'Indented' : 'Not indented';
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Turns the input value into its JSON equivalent (stringifies the value).
      `,
            infoBoxTitle: 'To JSON Node',
            contextMenuTitle: 'To JSON',
            group: ['Text'],
        };
    }
    async process(inputs) {
        const data = coerceType(inputs['data'], 'any');
        const toJson = this.data.indented ? JSON.stringify(data, null, 2) : JSON.stringify(data);
        return {
            ['json']: {
                type: 'string',
                value: toJson,
            },
        };
    }
}
export const toJsonNode = nodeDefinition(ToJsonNodeImpl, 'To JSON');
