// @ironclad/rivet-core/dist/esm/model/nodes/TrimChatMessagesNode.js
import {} from '../../model/NodeBase.js';
import { NodeImpl } from '../../model/NodeImpl.js';
import { nanoid } from 'nanoid/non-secure';
import {} from '../../index.js';
import { dedent } from 'ts-dedent';
import { nodeDefinition } from '../NodeDefinition.js';
import { coerceType } from '../../utils/coerceType.js';
import { getInputOrData } from '../../utils/index.js';
export class TrimChatMessagesNodeImpl extends NodeImpl {
    static create() {
        const chartNode = {
            type: 'trimChatMessages',
            title: 'Trim Chat Messages',
            id: nanoid(),
            visualData: {
                x: 0,
                y: 0,
                width: 200,
            },
            data: {
                maxTokenCount: 4096,
                removeFromBeginning: true,
            },
        };
        return chartNode;
    }
    getInputDefinitions() {
        const inputs = [
            {
                id: 'input',
                title: 'Input',
                dataType: 'chat-message[]',
            },
        ];
        if (this.data.useMaxTokenCountInput) {
            inputs.push({
                id: 'maxTokenCount',
                title: 'Max Token Count',
                dataType: 'number',
            });
        }
        if (this.data.useRemoveFromBeginningInput) {
            inputs.push({
                id: 'removeFromBeginning',
                title: 'Remove From Beginning',
                dataType: 'boolean',
            });
        }
        return inputs;
    }
    getOutputDefinitions() {
        return [
            {
                id: 'trimmed',
                title: 'Trimmed',
                dataType: 'chat-message[]',
            },
        ];
    }
    getEditors() {
        return [
            {
                type: 'number',
                label: 'Max Token Count',
                dataKey: 'maxTokenCount',
                useInputToggleDataKey: 'useMaxTokenCountInput',
            },
            {
                type: 'toggle',
                label: 'Remove From Beginning',
                dataKey: 'removeFromBeginning',
                useInputToggleDataKey: 'useRemoveFromBeginningInput',
            },
        ];
    }
    getBody() {
        return dedent `
      Max Token Count: ${this.data.useMaxTokenCountInput ? '(From Input)' : this.data.maxTokenCount}
      Remove From Beginning: ${this.data.useRemoveFromBeginningInput ? '(From Input)' : this.data.removeFromBeginning ? 'Yes' : 'No'}
    `;
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Takes an array of chat messages, and slices messages from the beginning or the end of the list until the total length of the messages is under the configured token length.

        Useful for setting up infinite message chains that stay under the LLM context limit.
      `,
            infoBoxTitle: 'Trim Chat Messages Node',
            contextMenuTitle: 'Trim Chat Messages',
            group: ['AI'],
        };
    }
    async process(inputs, context) {
        const input = coerceType(inputs['input'], 'chat-message[]');
        const maxTokenCount = getInputOrData(this.data, inputs, 'maxTokenCount', 'number');
        const removeFromBeginning = getInputOrData(this.data, inputs, 'removeFromBeginning', 'boolean');
        const trimmedMessages = [...input];
        const tokenizerInfo = {
            node: this.chartNode,
        };
        let tokenCount = await context.tokenizer.getTokenCountForMessages(trimmedMessages, undefined, tokenizerInfo);
        while (tokenCount > maxTokenCount) {
            if (removeFromBeginning) {
                trimmedMessages.shift();
            }
            else {
                trimmedMessages.pop();
            }
            tokenCount = await context.tokenizer.getTokenCountForMessages(trimmedMessages, undefined, tokenizerInfo);
        }
        return {
            ['trimmed']: {
                type: 'chat-message[]',
                value: trimmedMessages,
            },
        };
    }
}
export const trimChatMessagesNode = nodeDefinition(TrimChatMessagesNodeImpl, 'Trim Chat Messages');
