// @ironclad/rivet-core/dist/esm/model/nodes/WaitForEventNode.js
import {} from '../NodeBase.js';
import { NodeImpl } from '../NodeImpl.js';
import { nodeDefinition } from '../NodeDefinition.js';
import { nanoid } from 'nanoid/non-secure';
import {} from '../GraphProcessor.js';
import { coerceType } from '../../utils/coerceType.js';
import {} from '../ProcessContext.js';
import { dedent } from 'ts-dedent';
import {} from '../EditorDefinition.js';
import {} from '../NodeBodySpec.js';
export class WaitForEventNodeImpl extends NodeImpl {
    static create() {
        return {
            id: nanoid(),
            type: 'waitForEvent',
            title: 'Wait For Event',
            visualData: { x: 0, y: 0, width: 150 },
            data: {
                eventName: 'continue',
                useEventNameInput: false,
            },
        };
    }
    getInputDefinitions() {
        const inputDefinitions = [];
        if (this.chartNode.data.useEventNameInput) {
            inputDefinitions.push({
                id: 'eventName',
                title: 'Event Name',
                dataType: 'string',
            });
        }
        inputDefinitions.push({
            id: 'inputData',
            title: 'Data',
            dataType: 'any',
        });
        return inputDefinitions;
    }
    getOutputDefinitions() {
        return [
            {
                id: 'outputData',
                title: 'Data',
                dataType: 'any',
            },
            {
                id: 'eventData',
                title: 'Event Data',
                dataType: 'any',
            },
        ];
    }
    getEditors() {
        return [
            {
                type: 'string',
                label: 'Event Name',
                dataKey: 'eventName',
                useInputToggleDataKey: 'useEventNameInput',
            },
        ];
    }
    getBody() {
        return this.data.useEventNameInput ? '(Using Input)' : this.data.eventName;
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Waits for a specific event to be raised by a 'Raise Event' node or the host project. The event name can be configured.
      `,
            infoBoxTitle: 'Wait For Event Node',
            contextMenuTitle: 'Wait For Event',
            group: ['Advanced'],
        };
    }
    async process(inputs, context) {
        const eventName = this.chartNode.data.useEventNameInput
            ? coerceType(inputs['eventName'], 'string')
            : this.chartNode.data.eventName;
        const eventData = await context.waitEvent(eventName);
        return {
            ['outputData']: inputs['inputData'],
            ['eventData']: eventData,
        };
    }
}
export const waitForEventNode = nodeDefinition(WaitForEventNodeImpl, 'Wait For Event');
